File: src/1568.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   private:
    const vector<vector<int>> Directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    struct ArticulationPointInfo {
        bool hasArticulationPoint;
        int time;

        ArticulationPointInfo(bool hasArticulationPoint, int time)
            : hasArticulationPoint(hasArticulationPoint), time(time) {}
    };

    void findArticulationPoints(vector<vector<int>>& grid, int row, int col, vector<vector<int>>& discoveryTime,
                                vector<vector<int>>& lowestReachable, vector<vector<int>>& parentCell,
                                ArticulationPointInfo& apInfo) {
        int rowSize = grid.size();
        int colSize = grid[0].size();

        discoveryTime[row][col] = apInfo.time;
        apInfo.time++;
        lowestReachable[row][col] = discoveryTime[row][col];

        int children = 0;
        for (const auto& direction : Directions) {
            int x = row + direction[0];
            int y = col + direction[1];
            if ((x < 0) || (x >= rowSize)) {
                continue;
            } else if ((y < 0) || (y >= colSize)) {
                continue;
            } else if (grid[x][y] != 1) {
                continue;
            }

            if (discoveryTime[x][y] == -1) {
                children++;

                parentCell[x][y] = row * colSize + col;
                findArticulationPoints(grid, x, y, discoveryTime, lowestReachable, parentCell, apInfo);

                // Update lowest reachable time
                lowestReachable[row][col] = min(lowestReachable[row][col], lowestReachable[x][y]);
                // Check for articulation point condition
                if ((lowestReachable[x][y] >= discoveryTime[row][col]) && (parentCell[row][col] != -1)) {
                    apInfo.hasArticulationPoint = true;
                }
            } else if (x * colSize + y != parentCell[row][col]) {
                // Update lowest reachable time for back edge
                lowestReachable[row][col] = min(lowestReachable[row][col], discoveryTime[x][y]);
            }
        }

        // Root of DFS tree is an articulation point if it has more than one child
        if (parentCell[row][col] == -1 && children > 1) {
            apInfo.hasArticulationPoint = true;
        }
    }

   public:
    int minDays(vector<vector<int>>& grid) {
        int retVal = 2;

        int rowSize = grid.size();
        int colSize = grid[0].size();

        // Time when a cell is first discovered
        vector<vector<int>> discoveryTime(rowSize, vector<int>(colSize, -1));
        // Lowest discovery time reachable from the subtree rooted at this cell
        vector<vector<int>> lowestReachable(rowSize, vector<int>(colSize, -1));
        // Parent of each cell in DFS tree
        vector<vector<int>> parentCell(rowSize, vector<int>(colSize, -1));

        ArticulationPointInfo apInfo(false, 0);
        int landCells = 0;
        int islandCount = 0;
        for (int x = 0; x < rowSize; ++x) {
            for (int y = 0; y < colSize; ++y) {
                if (grid[x][y] != 1) {
                    continue;
                }
                landCells++;

                if (discoveryTime[x][y] != -1) {
                    continue;
                }
                findArticulationPoints(grid, x, y, discoveryTime, lowestReachable, parentCell, apInfo);
                islandCount++;
            }
        }
        if ((islandCount == 0) || (islandCount >= 2)) {
            retVal = 0;  // Already disconnected or no land
        } else if (landCells == 1) {
            retVal = 1;  // Only one land cell
        } else if (apInfo.hasArticulationPoint) {
            retVal = 1;  // An articulation point exists
        }

        return retVal;  // Need to remove any two land cells
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<vector<int>> grid;
    };
    vector<subject> testData{{{{0, 1, 1, 0}, {0, 1, 1, 0}, {0, 0, 0, 0}}}, {{{1, 1}}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: grid = [[0,1,1,0],[0,1,1,0],[0,0,0,0]]
     *  Output: 2
     *
     *  Input: grid = [[1,1]]
     *  Output: 2
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: grid = [";
        for (long unsigned int j = 0; j < testData[i].grid.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << "[";
            for (long unsigned int k = 0; k < testData[i].grid[j].size(); ++k) {
                cout << ((k == 0) ? "" : ",") << testData[i].grid[j][k];
            }
            cout << "]";
        }
        cout << "]\n";

        answer = cSolution.minDays(testData[i].grid);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/1568.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def __init__(self) -> None:
        self.Directions = [(0, 1), (0, -1), (1, 0), (-1, 0),]

    class ArticulationPointInfo:
        def __init__(self, hasArticulationPoint, time):
            self.hasArticulationPoint = hasArticulationPoint
            self.time = time

    def findArticulationPoints(self, grid: List[List[int]], row: int, col: int,
                               discoveryTime: List[List[int]],
                               lowestReachable: List[List[int]],
                               parentCell: List[List[int]],
                               apInfo: ArticulationPointInfo):
        rowSize = len(grid)
        colSize = len(grid[0])

        discoveryTime[row][col] = apInfo.time
        apInfo.time += 1
        lowestReachable[row][col] = discoveryTime[row][col]

        children = 0
        for direction in self.Directions:
            x = row + direction[0]
            y = col + direction[1]
            if (x < 0) or (x >= rowSize):
                continue
            elif (y < 0) or (y >= colSize):
                continue
            elif grid[x][y] != 1:
                continue

            if discoveryTime[x][y] == -1:
                children += 1

                parentCell[x][y] = row * colSize + col
                self.findArticulationPoints(grid, x, y, discoveryTime, lowestReachable, parentCell, apInfo)

                # Update lowest reachable time
                lowestReachable[row][col] = min(lowestReachable[row][col], lowestReachable[x][y])
                # Check for articulation point condition
                if (lowestReachable[x][y] >= discoveryTime[row][col]) and (parentCell[row][col] != -1):
                    apInfo.hasArticulationPoint = True
            elif x * colSize + y != parentCell[row][col]:
                # Update lowest reachable time for back edge
                lowestReachable[row][col] = min(lowestReachable[row][col], discoveryTime[x][y])

        # Root of DFS tree is an articulation point if it has more than one child
        if (parentCell[row][col] == -1) and (children > 1):
            apInfo.hasArticulationPoint = True

    def minDays(self, grid: List[List[int]]) -> int:
        retVal = 2

        rowSize = len(grid)
        colSize = len(grid[0])

        # Time when a cell is first discovered
        discoveryTime = [[-1] * colSize for _ in range(rowSize)]
        # Lowest discovery time reachable from the subtree rooted at this cell
        lowestReachable = [[-1] * colSize for _ in range(rowSize)]
        # Parent of each cell in DFS tree
        parentCell = [[-1] * colSize for _ in range(rowSize)]

        apInfo = self.ArticulationPointInfo(False, 0)
        landCells = 0
        islandCount = 0
        for x in range(rowSize):
            for y in range(colSize):
                if grid[x][y] != 1:
                    continue
                landCells += 1

                if discoveryTime[x][y] != -1:
                    continue
                self.findArticulationPoints(grid, x, y, discoveryTime, lowestReachable, parentCell, apInfo)
                islandCount += 1

        if islandCount == 0 or islandCount >= 2:
            retVal = 0  # Already disconnected or no land
        elif landCells == 1:
            retVal = 1  # Only one land cell
        elif apInfo.hasArticulationPoint:
            retVal = 1  # An articulation point exists

        return retVal  # Need to remove any two land cells


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for grid in [[[0, 1, 1, 0], [0, 1, 1, 0], [0, 0, 0, 0]], [[1, 1]]]:
            # /* Example
            #  *  Input: grid = [[0,1,1,0],[0,1,1,0],[0,0,0,0]]
            #  *  Output: 2
            #  *
            #  *  Input: grid = [[1,1]]
            #  *  Output: 2
            #  */
            logging.info("Input: grid = %s", grid)

            retVal = pSolution.minDays(grid)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
